sqlmesh/core/metric/definition.py

```diff
@@ -0,0 +1,193 @@
+
+
+def remove_namespace(expression: str | exp.Column, dialect: t.Optional[str] = None) -> str:
+    """Given a column or a string, rewrite table namespaces like catalog.db to catalog__db"""
+
+    if not isinstance(expression, str):
+        assert dialect is not None
+        expression = first(find_tables(expression, dialect=dialect))
+    return expression.replace(".", "__")
+
+
+class MetricMeta(PydanticModel, frozen=True):
+    """Raw metric definition without relationships or expansion of derived metrics."""
+
+    name: str
+    dialect: str
+    expression: exp.Expression
+    description: t.Optional[str] = None
+    owner: t.Optional[str] = None
+
+    _path: Path = Path()
+
+    @validator("name", pre=True)
+    def _name_validator(cls, v: t.Any, values: t.Dict[str, t.Any]) -> str:
+        return cls._string_validator(v).lower()
+
+    @validator("dialect", "owner", "description", pre=True)
+    def _string_validator(cls, v: t.Any) -> t.Optional[str]:
+        if isinstance(v, exp.Expression):
+            return v.name
+        return str(v) if v is not None else None
+
+    @validator("expression", pre=True)
+    def _validate_expression(
+        cls,
+        v: t.Any,
+        values: t.Dict[str, t.Any],
+    ) -> exp.Expression:
+        if isinstance(v, str):
+            return d.parse_one(v, dialect=values.get("dialect"))
+        if isinstance(v, exp.Expression):
+            return v
+        return v
+
+    def to_metric(
+        self, metas: t.Dict[str, MetricMeta], metrics: UniqueKeyDict[str, Metric]
+    ) -> Metric:
+        """Converts a metric meta into a fully expanded and standalone metric."""
+        payload = self.dict(exclude={"expression"})
+        metric_refs = {}
+
+        agg_or_ref = False
+
+        for node, *_ in self.expression.walk():
+            if isinstance(node, exp.Alias):
+                _raise_metric_config_error(
+                    f"Alias found for metric '{self.name}' which is not allowed", self._path
+                )
+            elif isinstance(node, exp.AggFunc):
+                agg_or_ref = True
+            elif isinstance(node, exp.Column) and not node.table:
+                agg_or_ref = True
+                ref = d.normalize_model_name(node.sql(dialect=self.dialect), dialect=self.dialect)
+
+                if ref not in metrics:
+                    metrics[ref] = metas[ref].to_metric(metas, metrics)
+
+                metric_refs[node] = metrics[ref]
+
+        if not agg_or_ref:
+            _raise_metric_config_error(
+                f"Metric '{self.name}' missing an aggregation or metric ref", self._path
+            )
+
+        if metric_refs:
+            expanded = self.expression.copy()
+            for column in expanded.find_all(exp.Column):
+                metric = metric_refs.get(column)
+
+                if metric:
+                    column.replace(metric.expanded.copy())
+        else:
+            expanded = exp.alias_(self.expression, self.name)
```

Comment:
Why don't we alias this metric's expression in case when `metric_refs` is not empty? Doesn't this break multiple levels of nestedness? Eg. Metric A -> Metric B -> Metric C

Reply:
look at the test cases here, but we don't need the aliases on the derived metrics

Reply:
you'd probably get something like

sum(x) as a / sum(b as ) b as c

Reply:
they are for atomic metrics